### DISCUSSION TITLE
Parse query file, how to extract query names and query text

### DISCUSSION THREAD

**Discussion by tobiaseriksson** [graphql-java/graphql-java#2518] [Q&A]
I have a set of files with graphql queries, and I would like to parse/validate them and create a map so that I can then pass them on to the graphql service

Pretty much something like this :
1. my-query-file.graphql
2. parse file (and validate)
3. Map<String,String> queries such that Key = name of query and Value = Text representation of the full query
4. Have a call that specifies the "key" i.e. query name and send this query to the graphql service

I have found out that I can use the 
grapqhl.parser.Parser to parse the queries into a Document and from there there I can see the Definitions
But the Definitions I can't easily convert back to the text representation of the query so that I can send it off and be executed 
Any ideas on how I could do this ?
-Tobias

**Comment by bbakerman** [comment]
When you say you want to parse and validate them but keep them in text form (Map<String,String> you know you will still incur  the parsing and validation per request right.  Its fast but it always converts string input to a full AST.

You have a couple of mechanisms

* `graphql.ParseAndValidate` allows you to parse  and validate some query input.  If its valid, just hold onto the original query string.
* Nominally from an AST Document you can use `graphql.language.AstPrinter` to get back to a String format.

More complicated is `graphql.execution.preparsed.PreparsedDocumentProvider` which allows you to "cache" parsed and validated documents in memory - this interface is called back to ask for a Document, which you may have previously loaded once say - its can also act as a pull through cache.

**Reply by tobiaseriksson** [reply]
This is exactly what I was after the AstPrinter was what I was after
Thanx, much appreciated !
